redesign: Add a new personal menu.

Ddharmani3 opened the pull request:
<!-- Describe your pull request here.-->
Added a new personal menu dropdown as a part of the #22802 redesign.

Original work on this is done in #24140 by @Taiki-Choda.
I tried rebasing #24140 first but since it was last updated in January, there were a lot of errors.
So, I created a new branch and took the work done in #24140 as per requirements.
Most of the work is the same, the changes I did are mostly in CSS.

Follow-up PRs:
Theme switcher #22803(I will open a PR once this PR gets merged)
Redesign the gear menu(I will open a PR for this soon).
Redesign user cards. (I will open a PR after this PR and #25459 get merged).


Fixes part of #22802.
Fixes: <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![image](https://github.com/zulip/zulip/assets/64723994/f18323a7-90f2-4627-a584-15b3c980161d)

![image](https://github.com/zulip/zulip/assets/64723994/520caf7a-7459-4cfb-b3ba-20cfe9e6e2d0)

![image](https://github.com/zulip/zulip/assets/64723994/0dcb251a-2394-40e8-96cb-028a17c914a8)

![image](https://github.com/zulip/zulip/assets/64723994/68f12afe-fd96-421f-b374-630bc9acab0a)

![image](https://github.com/zulip/zulip/assets/64723994/a99e3ec2-6cb2-424c-9922-c05f53588b81)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Ddharmani3 commented:
Feedback required:
How the user name in the dropdown should break? ( word-break: break-word; vs word-break: break-all;)
Example user name: verylongfirstname verylongmiddlename verylonglastname
  -  word-break: break-word;(currently done in this PR)
  ![image](https://github.com/zulip/zulip/assets/64723994/058932cf-1303-487f-8936-4f11d4916b0d)

  -  word-break: break-all; 
  ![image](https://github.com/zulip/zulip/assets/64723994/88cc8142-57c3-49f4-a24b-ac6f0f7165d1)

Ddharmani3 commented:
@alya this PR is ready for review now.

alya commented:
@amanagr Before I spend time reviewing this PR, did we want to split parts of it out? I recall that the plan was not to try to integrate all the changes in one fell swoop.

Ddharmani3 commented:
> @amanagr Before I spend time reviewing this PR, did we want to split parts of it out? I recall that the plan was not to try to integrate all the changes in one fell swoop.

@alya, I have split the redesign changes in #22802 into multiple PRs as per the previously discussed plan(also mentioned in PR description):

- This PR only adds a new personal and which can be reviewed separately, although as suggested [here](https://chat.zulip.org/#narrow/stream/9-issues/topic/log.20out.20not.20visible/near/1572720) I will convert this menu to tippy soon, it will not affect the UI, but I think this can be reviewed after I migrate it to tippy, since this PR is not going to be merged before 7.0 release.
- The next PR will be the gear menu redesign, I have completed work on this but as per the suggestion mentioned above I will convert this menu to tippy and open the PR soon.
- The last PR will be to redesign the user card, which can be opened once this PR and #25459 get merged.

amanagr commented:
1st commit: 
Only add SVGs that are being used here.
2nd commit: 
There are a lot unrelated changes in there. Let's split this into separate commits so that it is easy to understand why some CSS changes are required. (Like I don't understand why we are touching spectator CSS or background color).

I feel like the amount of constant numbers we are using for right navbar buttons is unhealthy. Would be up for a separate PR to move these items to a flexbox?

shameondev commented:
Hey, @Ddharmani3. I'm adding the "QA Needed" label that we discussed at [CZO](https://chat.zulip.org/#narrow/stream/2-general/topic/.22QA.20needed.22.20PR.20label), so that other folks who are interested can test your work. Feel free to add it yourself next time if you think it makes sense!

ryanreh99 commented on web/templates/personal_menu.hbs:
```diff
@@ -0,0 +1,157 @@
+        </header>
+        <section class="popover-menu info_popover_actions" data-user-id="{{user_id}}">
+            <ul class="list">
+                {{#if status_content_available}}
+                <li class="text-item">
+                    <span>
+                        {{#if status_emoji_info}}
+                        {{#if status_emoji_info.emoji_alt_code}}
+                        <span class="emoji_alt_code">&nbsp:{{status_emoji_info.emoji_name}}:</span>
+                        {{else if status_emoji_info.url}}
+                        <img src="{{status_emoji_info.url}}" class="emoji status_emoji" />
+                        {{else}}
+                        <span class="emoji status_emoji emoji-{{status_emoji_info.emoji_code}}"></span>
+                        {{/if}}
+                        {{/if}}
+                        <span class="status_text">
+                            {{status_text}}
+                        </span>
+                    </span>
+                    <a href="" class="clear_status" aria-label="{{t 'Clear status'}}" data-tippy-content="{{t 'Clear your status' }}">
+                        <i class="zulip-icon zulip-icon-x-circle"></i>
+                    </a>
+                </li>
+                <li class="link-item hidden-for-spectators">
+                    <a href="" class="update_status_text">
+                        <i class="zulip-icon zulip-icon-smile"></i>
+                        {{#tr}}Edit status{{/tr}}
+                    </a>
+                </li>
```
Since we are updating this in the user card migration #25459 as well.
Might want to split this out in another handlebar inorder to not duplicate code.
https://github.com/zulip/zulip/pull/25459/files#diff-d1c1238c4dc94e7f202b56433ac6b92f8a3f4b7331a54283ac4ae26cb21645d8R77-R98

btw missing semi-colon in `&nbsp;`

ryanreh99 replied:
and `hidden-for-spectator` class is not required right ?

Ddharmani3 replied:
> Since we are updating this in the user card migration #25459 as well. Might want to split this out in another handlebar inorder to not duplicate code. https://github.com/zulip/zulip/pull/25459/files#diff-d1c1238c4dc94e7f202b56433ac6b92f8a3f4b7331a54283ac4ae26cb21645d8R77-R98
> 
> btw missing semi-colon in `&nbsp;`

I will fix the semi-colon in the updated PR. 

For the split-out part, you suggested I am thinking of doing it in user card redesign PR as not only this but almost all of the design is the same in the profile menu and user card, you can check this is #22802. Thoughts on this plan?

Ddharmani3 replied:
Fixed both Issues and also fixed the dropdown disappearing on resize in updated PR.

ryanreh99 commented on web/src/popover_menus.js:
```diff
@@ -918,4 +929,106 @@ export function initialize() {
+            $popper.one("click", ".clear_status", (e) => {
+                e.preventDefault();
+                const me = page_params.user_id;
+                user_status.server_update_status({
```
lets create a function `clear_status`, as its called twice now.

Ddharmani3 replied:
@ryanreh99 I am not sure what you are trying to say here I was unable to find any code repetition here. Are you mentioning that there are 2 click handlers for `.clear_status`?

Can you please clarify more on this feedback?

ryanreh99 reviewed (commented):
added 2 refactoring comments.

I see similar refactoring changes here (`web/src/popovers.js`) as mentioned in the gear menu PR #25732
lets keep a note of it while handling merge conflicts.

Ddharmani3 commented:
> added 2 refactoring comments.
> 
> I see similar refactoring changes here (`web/src/popovers.js`) as mentioned in the gear menu PR #25732 let's keep a note of it while handling merge conflicts.

Yes, I am aware that a lot of things are common and I have kept the structure of these 2 PRs mostly the same you can also compare that a lot of commits are similar even the commit messages are exactly the same. 

And I plan to resolve conflicts in gear menu PR #25732 after this PR gets merged, I have also shared this plan [here](https://github.com/zulip/zulip/pull/25732#issuecomment-1572736000).

Ddharmani3 commented:
Resolved the conflicts and improved CSS slightly.
This PR is ready for review now.

Ddharmani3 commented:
Added `?` hotkey to the profile menu as well as suggested for the gear menu([here](https://github.com/zulip/zulip/pull/25732#issuecomment-1601268856)):
![image](https://github.com/zulip/zulip/assets/64723994/13d4261e-2e46-4c20-916a-ab6e031dc60b)
![image](https://github.com/zulip/zulip/assets/64723994/d0386064-2659-4f24-a558-a5ef9c954b88)

amanagr commented on web/templates/navbar.hbs:
```diff
@@ -40,17 +40,23 @@
-            <div id="navbar-buttons" {{#if embedded}}class="hide-navbar-buttons-visibility"{{/if}}>
+            <div id="navbar-buttons" class="header-button {{#if embedded}}hide-navbar-buttons-visibility{{/if}}">
+            </div>
+            <div id="personal-menu" class="dropdown actual-dropdown-menu hidden-for-spectators">
```
We are we using `dropdown` and `actual-dropdown-menu` classes here?

amanagr commented on web/templates/navbar.hbs:
```diff
@@ -40,17 +40,23 @@
+            <div id="navbar-buttons" class="header-button {{#if embedded}}hide-navbar-buttons-visibility{{/if}}">
+            </div>
+            <div id="personal-menu" class="dropdown actual-dropdown-menu hidden-for-spectators">
+                <a class="header-button dropdown-toggle tippy-zulip-delayed-tooltip" tabindex="0" role="button" data-target="nada" data-toggle="dropdown" data-tippy-content="{{t 'Personal menu' }}">
```
Similarly `dropdown-toggle` and related attributes.

amanagr commented:
Should we show an icon while user avatar is loading?  Otherwise it looks like there is nothing there.
<img width="237" alt="image" src="https://github.com/zulip/zulip/assets/25124304/8f2fc8fc-9063-482b-9d24-a291437825e3">

amanagr commented on web/src/popover_menus.js:
```diff
@@ -918,4 +925,89 @@ export function initialize() {
+            instance.setContent(
+                parse_html(
+                    render_personal_menu({
+                        user_id: my_user_id,
+                        invisible_mode,
+                        user_is_guest: page_params.is_guest,
+                        spectator_view: page_params.is_spectator,
+
+                        // narrow urls
+                        sent_by_url: hash_util.by_sender_url(my_email),
+                        pm_with_url: hash_util.pm_with_url(my_email),
+
+                        // user information
+                        user_avatar: page_params.avatar_url_medium,
+                        is_active: people.is_active_user_for_popover(my_user_id),
+                        user_circle_class: buddy_data.get_user_circle_class(my_user_id),
+                        user_last_seen_time_status:
+                            buddy_data.user_last_seen_time_status(my_user_id),
+                        user_full_name: page_params.full_name,
+                        user_type: people.get_user_type(my_user_id),
+
+                        // user status
+                        status_content_available: Boolean(status_text || status_emoji_info),
+                        status_text,
+                        status_emoji_info,
+                        user_time: people.get_user_time(my_user_id),
```
Let's move this to `popover_menus_data`.

amanagr commented:
Keyboard navigation doesn't go to `Logout` button, is that intentional?

amanagr commented on web/src/popover_menus.js:
```diff
@@ -933,6 +933,17 @@ export function initialize() {
     register_popover_menu("#personal-menu", {
         placement: "bottom",
         offset: [-50, 0],
+        popperOptions: {
+            strategy: "fixed",
+            modifiers: [
+                {
+                    name: "eventListeners",
+                    options: {
+                        scroll: false,
+                    },
+                },
+            ],
+        },
```
Let's add a comment in the code for this as well.

amanagr replied:
Let's add a comment in the code.

amanagr commented on web/templates/gear_menu.hbs:
```diff
@@ -176,7 +176,7 @@
                 </a>
             </li>
             <li role="presentation">
-                <a href="#logout" class="logout_button hidden-for-spectators" role="menuitem">
+                <a class="logout_button hidden-for-spectators" role="menuitem">
                     <i class="fa fa-power-off" aria-hidden="true"></i> {{t 'Log out' }}
                 </a>
```
Let's also do this for gear menu? (In a separate PR perhaps)

amanagr commented:
@Ddharmani3 looks great! Left a few comments.

Ddharmani3 commented:
> Should we show an icon while user avatar is loading? Otherwise it looks like there is nothing there before the user avatar appears. <img alt="image" width="237" src="https://user-images.githubusercontent.com/25124304/251335811-8f2fc8fc-9063-482b-9d24-a291437825e3.png">

Should I use the small avatar instead of a medium and replace div with <img> here: `<div class="header-button-avatar" style="background-image: url('{{user_avatar}}');"></div>` and set user_avatar as src to avoid this state?

@amanagr addresses all the other feedback.

amanagr commented:
I think that's a question for #design thread. Let's start a discussion for this there.

amanagr commented:
Please link the design thread here when you open one.

Ddharmani3 commented:
CZO: https://chat.zulip.org/#narrow/stream/101-design/topic/User.20avatar.20in.20navbar/near/1607247

amanagr commented:
@Ddharmani3 let me know when this is ready for review.

Ddharmani3 commented:
> @Ddharmani3 let me know when this is ready for review.

@amanagr, This is ready for review. I think the checks have failed because there was some issue with github yesterday when I pushed the updated code; the issue was the same as another PR, too. I am re running the check, but other than that, there is no change in code, and PR is ready for review :)

amanagr commented:
We should close gear menu when personal menus is open and vise versa.
<img width="288" alt="image" src="https://github.com/zulip/zulip/assets/25124304/95264c07-a1a2-4de0-a10d-5252b10c397f">

Ddharmani3 commented:
> We should close gear menu when personal menus is open and vise versa.
> <img width="288" alt="image" src="https://github.com/zulip/zulip/assets/25124304/95264c07-a1a2-4de0-a10d-5252b10c397f">
> 

After gear menu redesign PR get merged, the behavior will be the same as you mentioned. As both are tippy and have hideonclick: true. so when clicked on gear icon persnol menu will close and vice versa.

amanagr commented on web/templates/navbar.hbs:
```diff
@@ -40,17 +40,23 @@
+            <div id="navbar-buttons" class="header-button {{#if embedded}}hide-navbar-buttons-visibility{{/if}}">
+            </div>
+            <div id="personal-menu" class="hidden-for-spectators">
+                <a class="header-button tippy-zulip-delayed-tooltip" tabindex="0" role="button" data-target="nada" data-toggle="dropdown" data-tippy-content="{{t 'Personal menu' }}">
```
We don't need `data-toggle` and `data-target` here.

amanagr commented on web/styles/zulip.css:
```diff
@@ -611,8 +611,9 @@ body.has-overlay-scrollbar {
         justify-content: space-evenly;
```
Let's squash CSS changes in 2nd, 3rd, 4th and 5th commit together since it doesn't make sense to keep them apart to try to make sense.

amanagr commented on web/templates/personal_menu.hbs:
```diff
@@ -0,0 +1,148 @@
+                {{/if}}
+            </ul>
+
+            <ul class="list">
```
Let's just use one `ul` tag here.

amanagr replied:
So, I meant we should use one `ul` for the whole list and not multiple `ul` for each segment. We can use `.divider` elements as they are present in gear menu.

amanagr commented on web/src/popover_menus.js:
```diff
@@ -967,6 +967,11 @@ export function initialize() {
 
+            $(".focus-dropdown").on("focus", (e) => {
+                e.preventDefault();
+                $popper.find("li:not(.divider):visible a").eq(0).trigger("focus");
```
I don't see `.divider` class being used.

amanagr replied:
You can add this back based on my comment above https://github.com/zulip/zulip/pull/25667#discussion_r1266453935

amanagr commented on web/templates/navbar.hbs:
```diff
@@ -43,6 +43,7 @@
                     <a class="header-button login_button">
                         <i class="fa fa-sign-in"></i>
                     </a>
+                    <span tabindex="0" class="focus-dropdown"></span>
                 </div>
```
Can you explain why do we have a `focus-dropdown` class here?

amanagr commented on web/src/popover_menus.js:
```diff
@@ -967,6 +967,11 @@ export function initialize() {
                 });
             });
 
+            $(".focus-dropdown").on("focus", (e) => {
```
Why do we have a `focus-dropdown` handler inside `onMount`? I basically don't understand this commit.

amanagr replied:
In that case, we should make this event handler specific to this popover and not add global event handler. Use `$popper`

amanagr commented on web/src/popover_menus.js:
```diff
@@ -921,4 +925,57 @@ export function initialize() {
+            });
+
+            instance.popperInstance.update();
+        },
```
There is supposed to be only one `clear_status` button right?

amanagr commented on web/src/popovers.js:
```diff
@@ -1083,8 +1083,15 @@ export function any_active() {
     if (!opts || !opts.not_hide_tippy_instances) {
-        // hideAll hides all tippy instances (tooltips and popovers).
-        hideAll();
+        if (opts && opts.not_hide_tippy_dropdowns) {
+            // don't hide gear menu and personal menu dropdows
+            hideAll({
+                exclude: popover_menus.get_personal_menu_instance(),
+            });
```
We can just check `opts.not_hide_tippy_dropdowns` here since `opts` is defined. I think this if fine for now but let's add a comment that when we add "gear menu" then we should add a function which gets the current open tippy dropdown for us which is just either personal menu or gear menu.

amanagr commented on web/styles/zulip.css:
```diff
@@ -3311,3 +3311,12 @@ select.invite-as {
     }
 }
+
+.keyboard-shortcut-option {
+    & span.tooltip-hotkey-hint {
+        margin-left: auto;
+        padding: 0 4px;
+        color: hsl(227deg 78% 59%);
+        border-color: hsl(227deg 78% 59%);
+    }
+}
```
Let's exract these color values so that we don't need to touch `dark_theme.css` for this.

amanagr reviewed (commented):
@Ddharmani3 did an extensive review. I think once these are resolved. It will ready for `integration review`.

Ddharmani3 commented:
@amanagr addressed all the feedback. This PR is ready for review now.
Also, one question should I add the last commit in this PR or gear menu PR?

amanagr commented:
@Ddharmani3 thanks for the update. Replied to some old comments.

amanagr commented:
@Ddharmani3 looks good to me.

alya commented:
Testing this PR, I noticed that if you click on the gear menu and then the avatar menu, the gear menu incorrectly remains open:
<img width="342" alt="Screen Shot 2023-09-21 at 10 16 35 AM" src="https://github.com/zulip/zulip/assets/2090066/9e2ba7ea-75be-4521-bf2c-182973205777">

alya commented:
Looking at #22802, the time in @terpimost 's designs is above the status section in the menu. Was there discussion elsewhere about changing that? It looks out of place to me in its current location.


<img width="309" alt="Screen Shot 2023-09-21 at 10 25 14 AM" src="https://github.com/zulip/zulip/assets/2090066/2e1fc8b4-3e6d-4363-80f3-f17730489b0a">

alya commented:
That's all the feedback I have from playing with the PR, so from a user perspective, I think this is getting close!

Ddharmani3 commented:
> Testing this PR, I noticed that if you click on the gear menu and then the avatar menu, the gear menu incorrectly remains open:

This will be fixed when the gear menu migrates to tippy after merging #25732, this was [Aman's concern too](https://github.com/zulip/zulip/pull/25667#issuecomment-1632569132) :)

Ddharmani3 commented:
> Looking at #22802, the time in @terpimost 's designs is above the status section in the menu. Was there discussion elsewhere about changing that? It looks out of place to me in its current location.

@alya, The last version of @Taiki-Choda's PR had this version, I don't remember reading any discussion about it.

It was looking out of place to me too, changed it back to @terpimost's original design :)

Also updated the PR to resolve conflicts. Can you take another look at the PR?

timabbott commented on web/src/user_card_popover.js:
```diff
@@ -648,7 +648,7 @@ function register_click_handlers() {
+        const user_id = elem_to_user_id($(e.target).closest(".user-card-popover-actions"));
         const email = people.get_by_user_id(user_id).email;
         hide_all();
         if (overlays.is_active()) {
```
I'd rather we just do new click handlers rather than the reuse here; then the `overlays.is_active` check isn't necessary here, for example, and it's easier to verify things if independent components just have their own click handlers.

timabbott replied:
Fixed.

timabbott commented on web/src/popover_menus.js:
```diff
@@ -1180,6 +1180,11 @@ export function initialize() {
+            $(".focus-dropdown").on("focus", (e) => {
+                e.preventDefault();
+                $popper.find("li:visible a").eq(0).trigger("focus");
+            });
```
I don't quite understand what you're doing here, but I would expect keyboard UI for this menu to use the `popover_items_handle_keyboard` code that all of our other popovers use -- what's the thinking about that?

(I don't think `tab` behavior is important -- just match what we do with arrow keys in normal popovers)

timabbott replied:
This question remains open; could be a follow-up.

timabbott commented on web/src/popovers.js:
```diff
@@ -31,8 +31,17 @@ export function any_active() {
+        } else {
+            // hideAll hides all tippy instances (tooltips and popovers).
+            hideAll();
+        }
```
Let's drop this commit and leave this question for follow-ups; whether this new menu hides or not on scroll is not important, and https://github.com/zulip/zulip/pull/26884 is going to merge conflict from this and might happen to make this work the way we want it anyway.

timabbott replied:
Done

timabbott commented on web/src/popover_menus.js:
```diff
@@ -1159,6 +1159,19 @@ export function initialize() {
+        // The stratergy: "fixed"; and eventlistener modifier option ensures that the personal menu does not
+        // modify its position or disappear when user zoom the page.
+        popperOptions: {
+            strategy: "fixed",
```
This commit looks good except "strategy" is mispelled in the comment and the comment isn't line-wrapped properly. I think you should squash it into the main commit though.

timabbott replied:
Done

timabbott commented on web/templates/gear_menu.hbs:
```diff
@@ -176,7 +176,7 @@
             </li>
             <li role="presentation">
-                <a href="#logout" class="logout_button hidden-for-spectators" role="menuitem">
+                <a class="logout_button hidden-for-spectators" role="menuitem">
```
This probably should have been in its own PR, or a prep commit at the beginning, since it doesn't depend on the rest, but in any case, rebased to the start and merged as 55b0b5cb965b7dee850244baf55a4d96e5e465c1.

timabbott commented:
I rebased this past a bunch of `popover_menus.js` refactoring, addressed my feedback above, did some reordering in consultation with @alya, and removed the "help" cluster of items that are also in the gear menu to avoid duplication.

I think this needs a bit more testing for smaller screen size bugs -- there's just a lot of magic number of pixels changed. But possibly this is a version we can integrate.

timabbott commented:
Current screenshot (with a very long name for myself):

![image](https://github.com/zulip/zulip/assets/2746074/f9da1215-3c74-4909-9bd6-02551d6f1da8)

I think the font color for the user's role is a bit too low contrast, but that's probably an easy follow-up change.

alya commented:
A few comments from playing around with this:

1. The status dot for "offline" (in invisible mode) needs to be opaque, not transparent.
2. The tooltip on the status needs more space if you are invisible.
![Screenshot 2023-10-03 at 3 46 07 PM](https://github.com/zulip/zulip/assets/2090066/da770494-326a-4e82-a60f-be7aae8d24e9)

3. The issue with two menus being open at once noted above has not been resolved.
![Screenshot 2023-10-03 at 3 47 50 PM](https://github.com/zulip/zulip/assets/2090066/21fc70dc-a3f1-47f1-b164-bf28812f401c)

4. I think you planned to remove "Personal settings" from the gear menu?

alya commented:
@terpimost Would you like to do a review pass on this one?

timabbott commented:
> 1. The issue with two menus being open at once noted above has not been resolved.

I think this is best resolved as a follow-up in that it's a consequence of the gear menu not having been converted to a modern framework. But I added the fix for this because it was quick.


> 2. The status dot for "offline" (in invisible mode) needs to be opaque, not transparent.
> 3. The tooltip on the status needs more space if you are invisible.

@amanagr do you have time to make this fix?

> 4. I think you planned to remove "Personal settings" from the gear menu?
Done.

> 5. I think "Log out" can be dropped from the gear menu as well.

Done.

This is the state we end up with for the gear menu:

![image](https://github.com/zulip/zulip/assets/2746074/b1f7764e-39d6-4943-a0d0-ad40b0e1d14c)

Aside from all that, the changes to `right_sidebar.css` in the second commit are to me a basically unreviewable collection of random changes to the of pixels of various elements in the navbar area. I think they are not quite correct, e.g. in the very narrow view, it looks like the stream sidebar opener is a bit too high:

![image](https://github.com/zulip/zulip/assets/2746074/add93301-58ee-4ef8-8b46-14002d1f801a)

So we might need to test-deploy this for a bit or have someone pick up rewriting the CSS for that area to be less crazy (the problem is not new in this PR).

amanagr commented:
> @amanagr do you have time to make this fix?

Done.
<img width="267" alt="Screenshot 2023-10-04 at 9 44 52 AM" src="https://github.com/zulip/zulip/assets/25124304/7cd415d9-fd67-4313-be50-efa1cdd9b424">

alya commented:
Oh, sorry, when I said "opaque", I meant that just having the image peek through doesn't work, not that it should be white. It probably makes sense to match the "offline" dot color in the sidebars.


<img width="28" alt="Screenshot 2023-10-04 at 10 50 20 AM" src="https://github.com/zulip/zulip/assets/2090066/936f188b-60a2-4361-9a92-f25c9f4ed7d9">
<img width="24" alt="Screenshot 2023-10-04 at 10 50 11 AM" src="https://github.com/zulip/zulip/assets/2090066/b8de1834-8ac7-4626-bb8b-78e81ae141af">

timabbott commented:
Looks like a puppeteer failure; oh removing "settings" from the gear menu means we need to adjust the tests, OK.
```
TimeoutError: Waiting for selector `.dropdown-menu a[href^='#settings']` failed: Waiting failed: 30000ms exceeded
    at Timeout.<anonymous> (/__w/zulip/zulip/node_modules/.pnpm/puppeteer-core@21.2.1/node_modules/puppeteer-core/src/common/WaitTask.ts:92:11)
    at listOnTimeout (node:internal/timers:569:17)
    at processTimers (node:internal/timers:512:7)
```

terpimost commented:
The stroke around the circle should be the color of the current background:
<img width="45" alt="image" src="https://github.com/zulip/zulip/assets/1903309/c1c6aac5-ca45-4aca-b3bd-d4f1d4a37205">
In the dark theme it is a background of the current menu.. or at least lest to do the main BG of the app.

For the image I suggest to have a light border:
<img width="209" alt="image" src="https://github.com/zulip/zulip/assets/1903309/88047f91-62cf-4185-ae6b-6615afb7115c">

```
#personal-menu .header-button-avatar {
   
    border: 1px solid hsl(0deg 0% 0% / 10%);
/* for dar it could be border: 1px solid hsl(0deg 0% 100% / 20%);*/

}
```

I don't think that the text in the clickable menu items should be selectable:
<img width="265" alt="image" src="https://github.com/zulip/zulip/assets/1903309/5cc9b0d0-c56a-4b35-8c65-a9a3d92af122">

We shouldn't remove default browser outline on `:focus-visible`.

The way to do it is instead of this:


To have this (focus-visible instead of focus):
```
.popover-menu .link-item a:hover, .popover-menu .link-item a:focus-visible {
    background: var(--color-background-hover-dropdown-menu);
    /* outline: none; don't remove the outline */
    outline-offset: -2px;
}
```

But there are 2 problems of making a nice focus here:
1. There is a bootstrap style which takes over:
<img width="467" alt="image" src="https://github.com/zulip/zulip/assets/1903309/044acfcb-d952-495d-a8d7-9378e7418c5f">

2. There is a li and there is a inside it. I don't know which one should have focus: it could be solved like this:
<img width="493" alt="image" src="https://github.com/zulip/zulip/assets/1903309/91113321-ac63-4c8e-9c1b-a7474eac1991">

Ideally the result could be like this:
<img width="281" alt="image" src="https://github.com/zulip/zulip/assets/1903309/7c8ddeae-4fec-4c0f-b71e-8278c7ef1fde">

amanagr commented:
Updated based on all the feedback.
<img width="367" alt="Screenshot 2023-10-05 at 1 52 07 PM" src="https://github.com/zulip/zulip/assets/25124304/6816c160-f370-4931-8bd2-7df6b404096a">
<img width="367" alt="Screenshot 2023-10-05 at 1 51 58 PM" src="https://github.com/zulip/zulip/assets/25124304/a56dd843-b00c-4f23-b31a-516693bb60c3">
<img width="367" alt="Screenshot 2023-10-05 at 1 51 43 PM" src="https://github.com/zulip/zulip/assets/25124304/fd74347b-5fc8-41c6-9343-ea79e7620ac7">
<img width="367" alt="Screenshot 2023-10-05 at 1 51 36 PM" src="https://github.com/zulip/zulip/assets/25124304/b525a09c-e703-4f4a-abea-f14f79e622b2">

terpimost commented:
Sorry but we shouldn't do a smaller to get focus right:
<img width="380" alt="image" src="https://github.com/zulip/zulip/assets/1903309/2959d171-c13d-4760-9ccd-b8aed504e1eb">

We should control outline-offset. and default browser focus would have a proper roundings.
Right now the focus isn't browser's default and a item is smaller. Such menu items are possible but we are not going for those:
<img width="586" alt="image" src="https://github.com/zulip/zulip/assets/1903309/a8b0a7c5-5b5b-4f68-a241-1f90487b048e">

I'm not agains custom outlines but it is a big deal, so let use them only when we really need it.
I removed margin from `a` and `border-radius` and tuned the styles like that:
<img width="914" alt="image" src="https://github.com/zulip/zulip/assets/1903309/81a6ff0c-cdb2-43bb-9316-9541ae0f5db3">

There are places where explicit outline from links should be removed and focus should be replaced with focus-visible
<img width="669" alt="image" src="https://github.com/zulip/zulip/assets/1903309/279c42f9-30d5-4b9b-ac4f-ece80c1bb6bb">
<img width="681" alt="image" src="https://github.com/zulip/zulip/assets/1903309/bf0e7565-421d-4311-b03a-e41abbfb842b">

Correction: outline-offset could be just -1px. Result:

![CleanShot 2023-10-05 at 09 59 45](https://github.com/zulip/zulip/assets/1903309/a15687ba-70e2-4a5b-9738-a876827898dd)


Sorry but general background isn't good for stroke around the circle:
<img width="130" alt="image" src="https://github.com/zulip/zulip/assets/1903309/3371457d-08ab-4bbf-a4bc-6f3c2a06f9bf">
Lets figure out how can we control that... it could be a css variable on the parent level which would be redefined and by default it could be background of the app but in some cased we could override it like in menu and other places

Time isn't a clickable element but now its not possible to select it. It is not a big deal but only clickable elements shouldn't allow the selection.

I would like to suggest some design tune for hover and active elements of the menu:
**Light mode:**
hover bg:  `hsla(220, 12%, 5%, 0.05)`
active bg:  `hsla(220, 12%, 5%, 0.07)`

**Dark mode:**
hover bg: `hsla(220, 12%, 95%, 0.05)`
active bg: `hsla(220, 12%, 95%, 0.07)`

amanagr commented:
Ok, I modified it use default outline for the theme which was close to black for light theme and gray for dark theme.
<img width="261" alt="Screenshot 2023-10-06 at 1 21 36 PM" src="https://github.com/zulip/zulip/assets/25124304/11f5042b-cba3-4859-98b9-67d662eb71b0">


<img width="261" alt="Screenshot 2023-10-06 at 1 21 09 PM" src="https://github.com/zulip/zulip/assets/25124304/bf5ece0e-1183-4f6a-a5f5-a169733c6983">


Also, changed the background hover and active colors.

terpimost commented:
I'm sorry I meant outline of the status indicator not for the focus rectangle. Focus rectangle should be browser's default

timabbott commented:
This seems in a pretty good state. I note one bug:
* Using the keyboard and hitting `Enter` to visit "Settings" doesn't seem to work

Regarding the design for this particular popover:
* It feels pretty good to me. I am a little uncertain about the exact purple color we have for the icons; I worry a bit that it encourages one to just mentally filter out the icons. I'm not sure if that's a bad outcome? In any case, it's a single-CSS-variable detail so should definitely not delay the rest of this work, if we want to change it it all, but figured I'd mention it.
* I'm not sure about the icon for "messages sent"; posting it alongside the other "view messages" one since they also are going to appear next to each other in the user cards, which are a pretty high visibility surface area.
![image](https://github.com/zulip/zulip/assets/2746074/22419ae3-60e0-4237-af64-d9db3cc062cc)
here's what we have in `main`:
![image](https://github.com/zulip/zulip/assets/2746074/cf272ffa-dacd-49c7-8617-b6e9d384044f)
I guess mainly I liked the paper-airplane for "messages sent" as feeling a bit more different (not another rectangle) given the specific context that these are next to each other, but a chat bubble is not bad for "what this person has said".

amanagr commented:
> Using the keyboard and hitting Enter to visit "Settings" doesn't seem to work

Fixed.

timabbott commented on web/templates/personal_menu.hbs:
```diff
@@ -0,0 +1,135 @@
+                </li>
+                <li class="link-item">
+                    <a href="" class="update_status_text">
+                        <span class="link-item-focus">
```
Can you explain why we need this `link-item-focus` class to be a separate `span`? If we're going to be using these styles in all of our popovers eventually, it feels like a lot of potential.

Probably a thing to discuss in a thread in #frontend in case there's a way to avoid whatever constraint made you want to do this.

amanagr replied:
https://chat.zulip.org/#narrow/stream/6-frontend/topic/personal.20menus.20focus.20emulation

timabbott commented on web/src/click_handlers.js:
```diff
@@ -924,4 +924,10 @@ export function initialize() {
+        // Jquery prevents default action on anchor for `trigger("click")`
+        // so we need to use click on element to trigger the default action.
+        e.currentTarget.click();
+    });
```
Can you explain the motivation for this? I liked the previous model with explicit click handlers, both because it lets us set the `trigger` property (for logging) when adjusting the view and because that's the pattern we use on all of our other popovers, and this feels like a hacky workaround mechanism that, if it needed to exist, we'd have created quite some time ago.

Might be a good #frontend thread topic.

amanagr replied:
I realized we use it commonly in our code:

```
    if ($(".dropdown.open, .dropup.open").length > 0 && $(e.target).attr("role") === "menuitem") {
        // on dropdown menu elements, force a click and prevent default.
        // this is because these links do not have an href and so don't force a
        // default action.
        e.target.click();
        return true;
```

amanagr replied:
Although, the `role = menuitem` check is hard to guess what it does looking at the HTML code. So, we can migrate it to use `trigger-natural-click` class?

Let me start a frontend thread for this.

amanagr replied:
Ok, changed to use click handlers. We still need `e.currentTarget.click();` to open the settings othewise we will have to use `window.location.hash = ...` which we don't like to.

amanagr commented:
Updated the PR as per feedback from discussion in https://chat.zulip.org/#narrow/stream/6-frontend/topic/personal.20menus.20focus.20emulation

timabbott commented:
I made two last changes:
* Left "personal settings" in the gear menu -- I want to avoid disrupting workflows for finding settings until we've settled the gear menu redesign changes too. (Discussed with @alya).
* Renamed the misleading `.user-info-popover-header` to `personal-menu-header`.

Otherwise, this looks good to me, marked to merge once CI passes again.
